[iOS] Invoking date picker controls with large values may cause crash

lilyspiniolas opened the pull request:
#### 7ec0f38dfd7736be65b933384a700fecb9e73987
<pre>
[iOS] Invoking date picker controls with large values may cause crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=283317">https://bugs.webkit.org/show_bug.cgi?id=283317</a>
<a href="https://rdar.apple.com/135733092">rdar://135733092</a>

Reviewed by Wenson Hsieh.

An explicit date interval is now applied to the UIDatePicker
or UICalendarView used for date-related input types. If the value
of the input is outside of this interval, invoking the control
will set the value of the input to current day.

* LayoutTests/fast/forms/ios/date-picker-value-crash-expected.txt: Added.
* LayoutTests/fast/forms/ios/date-picker-value-crash.html: Added.
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker initWithView:inputType:]):
(-[WKDateTimePicker setWeekPickerToInitialValue]):
(-[WKDateTimePicker setDateTimePickerToInitialValue]):

Canonical link: <a href="https://commits.webkit.org/286777@main">https://commits.webkit.org/286777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1faeedd3bc92ebe16aa8ba7fa9aeaa6f1fefae31

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56061 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81575 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28305 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60363 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18434 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80093 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66117 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40674 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47715 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23615 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26631 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68833 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83010 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2957 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68636 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4561 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16940 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11872 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9956 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4352 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4372 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7807 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6131 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -85,6 +86,11 @@ - (id)initWithView:(WKContentView *)view inputType:(WebKit::InputType)inputType
     if (!(self = [super init]))
         return nil;
 
+    auto maximumDateFormatter = adoptNS([[NSDateFormatter alloc] init]);
```
```suggestion
    RetainPtr maximumDateFormatter = adoptNS([[NSDateFormatter alloc] init]);
```

aprotyas commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -85,6 +86,11 @@ - (id)initWithView:(WKContentView *)view inputType:(WebKit::InputType)inputType
 
+    auto maximumDateFormatter = adoptNS([[NSDateFormatter alloc] init]);
+    [maximumDateFormatter setDateFormat:kDateTimeFormatString];
+    auto maximumDate = adoptNS([maximumDateFormatter dateFromString: @"10000-12-31T23:59"]);
```
```suggestion
    RetainPtr maximumDate = adoptNS([maximumDateFormatter dateFromString: @"10000-12-31T23:59"]);
```

whsieh replied:
We generally prefer `RetainPtr … = adoptNS(…);` in new code.

```suggestion
    RetainPtr maximumDateFormatter = adoptNS([[NSDateFormatter alloc] init]);
    [maximumDateFormatter setDateFormat:kDateTimeFormatString];
    RetainPtr maximumDate = [maximumDateFormatter dateFromString:@"10000-12-31T23:59"];
```

The second `adoptNS(…)` will also cause an overrelease, since `+dateFromString:` does not return a +1 object.

rr-codes replied:
maybe add an inline comment here explaining the source of this value?

aprotyas commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -281,11 +290,13 @@ - (void)setWeekPickerToInitialValue
+    bool dateParsedAndSelectable = parsedDate && [[_calendarView availableDateRange] containsDate:parsedDate.get()];
 
-    RetainPtr<NSDateComponents> dateComponents = [[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601] components:unitFlags fromDate:parsedDate.get()];
+    RetainPtr<NSDateComponents> dateComponents = [[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601] components:unitFlags fromDate:dateParsedAndSelectable ? parsedDate.get() : [NSDate date]];
```
Nit: I know this is older code, but feel free to drop the template argument.
```suggestion
    RetainPtr dateComponents = [[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601] components:unitFlags fromDate:dateParsedAndSelectable ? parsedDate.get() : [NSDate date]];
```

whsieh reviewed (approved):
r=me (with comments addressed)

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -85,6 +86,11 @@ - (id)initWithView:(WKContentView *)view inputType:(WebKit::InputType)inputType
 
+    RetainPtr maximumDateFormatter = adoptNS([[NSDateFormatter alloc] init]);
+    [maximumDateFormatter setDateFormat:kDateTimeFormatString];
+    RetainPtr maximumDate = [maximumDateFormatter dateFromString: @"10000-12-31T23:59"]; // UIDatePicker cannot have more than 10,000 selectable years
```
Nit - still an extra space here

```suggestion
    RetainPtr maximumDate = [maximumDateFormatter dateFromString:@"10000-12-31T23:59"]; // UIDatePicker cannot have more than 10,000 selectable years
```

lilyspiniolas committed:
[iOS] Invoking date picker controls with large values may cause crash
https://bugs.webkit.org/show_bug.cgi?id=283317
rdar://135733092

Reviewed by Wenson Hsieh.

An explicit date interval is now applied to the UIDatePicker
or UICalendarView used for date-related input types. If the value
of the input is outside of this interval, invoking the control
will set the value of the input to current day.

* LayoutTests/fast/forms/ios/date-picker-value-crash-expected.txt: Added.
* LayoutTests/fast/forms/ios/date-picker-value-crash.html: Added.
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker initWithView:inputType:]):
(-[WKDateTimePicker setWeekPickerToInitialValue]):
(-[WKDateTimePicker setDateTimePickerToInitialValue]):

Canonical link: https://commits.webkit.org/286777@main
